h2o-py/tests/testdir_algos/automl/pyunit_automl_memory_cleanup.py

```diff
@@ -272,13 +292,192 @@ def test_param_disabled():
+        assert contains_event_log(project_name, keys)
+        expectations = dict(
+            models_base=max_models + 2,  # 2 SEs
+            cv_models=0,
+            predictions=0,
+            metrics=(max_models * 3  # for each non-SE model, 1 on training_frame, 1 on validation_frame, 1 on leaderboard_frame
+                     + (2 * 2)  # for each SE model, 1 on training frame, 1 on leaderboard frame
+                     + (2 * 4)  # for each SE metalearner, 1+1 on levelone training+validation, 1+1 for final scoring (preds_levelone training+leaderbird)
+                     )
+        )
+        for k, v in expectations.items():
+            assert len(keys[k]) == v, "expected {} {}, but got {}".format(v, k, len(keys[k]))
+
+        h2o.remove(aml)
+        clean = list_keys_in_memory()
+        # print(clean['all'].values)
+        assert not contains_leaderboard(project_name, clean)
+        assert not contains_event_log(project_name, clean)
+        assert len(clean['models_base']) == 0
+        assert len(clean['cv_models']) == 0
+        assert len(clean['models_all']) == 0
+        assert len(clean['predictions']) == 0
+        assert len(clean['metrics']) == 0
+        assert len(clean['automl']) == 0
+        for frame in [train, valid, test]:
+            assert frame_in_cluster(frame), "frame {} has been removed from cluster".format(frame.frame_id)
+
+
+    def test_remove_automl_with_xval_when_keeping_all_cv_details():
+        target, train, valid, test = prepare_data()
+        project_name = 'aml_with_xval_remove_test'
+        max_models = 3
+        nfolds = 5
+        aml = H2OAutoML(project_name=project_name,
+                        nfolds=nfolds,
+                        max_models=max_models,
+                        seed=1,
+                        keep_cross_validation_predictions=True,
+                        keep_cross_validation_fold_assignment=True,
+                        keep_cross_validation_models=True)
+        aml.train(y=target, training_frame=train)
+
+        keys = list_keys_in_memory()
+        # print(keys['all'].values)
+        assert contains_leaderboard(project_name, keys)
+        assert contains_event_log(project_name, keys)
+        expectations = dict(
+            models_base=max_models + 2,  # 2 SEs
+            cv_models=(max_models+2) * nfolds,  # 1 cv model per fold for all models, incl. SEs
+            predictions=(len(keys['cv_models'])  # cv predictions
+                         + len(keys['models_base'])  # cv holdout predictions
+                         ),
+            metrics=(len(keys['cv_models']) * 3  # for each cv model, 1 on training frame, 1 on validation frame (=training for cv), 1 on frame with null key (what is this?)
+                     + len(keys['models_base'])  # for each model, 1 on training_frame
+                     + (2 * 2)  # for each SE, 1 on levelone training, 1 for final scoring (preds_levelone)
```

Comment:
@michalkurka I'm quite happy with this test as it now gives me a – I expect complete – reference about all keys produced during AutoML and therefore for each model training.
For CV models although, I noticed there was 3 metrics stored in DKV for each of them: I can understand the one on training and validation frames, but a 3rd one is done against a frame with 'null' key. I spent some time searching in the code how it could happen but couldn't figure it out: any idea where this could come from? This looks like a temporary metric that should actually be deleted once the CV model is completed.
WDYT?

Reply:
that could be the metric calculated on the adapted training/validation frame - this frame doesn't have a key - it is strange though

are the metrics same as train or validation or completely different?

Reply:
@michalkurka Yep, just debugged this, and this are metrics computed on the adapted validation frame for CV holdout predictions.
It looks like this metrics key is stored as a byproduct of `Model#predictScoreImpl`, more exactly because we stored those metrics every time we set `computeMetrics=true`:
```
 if (computeMetrics && bs._mb != null) //metric builder can be null if training was interrupted/cancelled
      bs._mb.makeModelMetrics(this, fr, adaptFrm, bs.outputFrame());
```

It doesn't look necessary though in this case: tell me if I'm wrong but I don't see those metrics used anywhere.
Could create a small ticket to do some cleanup here.

Reply:
Thanks, I also don't see the reason to keep them there - please make the jira.